Add basic AVIF parsing support

baumanj opened the pull request:
Since the fundamentals of HEIF, upon which AVIF is based, are quite different
from video, this adds a new context type, which currently only contains the
data for the "primary item", typically an av01 coded image frame.

Extracting the primary item requires support for several new box types which
specify the location type of various data and metadata. Various features are
not yet fully supported, but should provide informative errors.

Since the specification does not make many guarantees about the ordering of
boxes, there is some additional logic complexity in an attempt to avoid
copying buffers when possible.

Also, a set of AVIF sample images (see the [ReadMe.txt](https://github.com/mozilla/mp4parse-rust/pull/193/files#diff-c737d664661ec1978a87d9209db32dd1) for licensing) is added
to exercise the new parsing code.

kinetiknz commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1173,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+    Ok(item_infos)
+}
+
+fn u32_to_string(src: u32) -> String {
```
be_u32_to_string or similar to make it clear it expects a big-endian u32?

kinetiknz commented on mp4parse/src/lib.rs:
```diff
@@ -821,6 +1033,40 @@ impl<'a, T: Read> BMFFBox<'a, T> {
+                if len > self.bytes_left() {
+                    return Err(Error::InvalidData("extent crosses box boundary"));
+                }
+                self.take(len).read_to_end(buf)?;
```
We'll need to make the extension of `buf` here by `read_to_end` use the fallible allocator, somehow.

kinetiknz commented on mp4parse/src/lib.rs:
```diff
@@ -821,6 +1033,40 @@ impl<'a, T: Read> BMFFBox<'a, T> {
+                self.take(len).read_to_end(buf)?;
+            }
+            ExtentRange::ToEnd(_) => {
+                self.read_to_end(buf)?;
```
As above.

kinetiknz commented on mp4parse/src/lib.rs:
```diff
@@ -689,6 +721,180 @@ impl MediaContext {
+            }
+        };
+        let slice = slice.ok_or(Error::InvalidData("extent crosses box boundary"))?;
+        buf.extend_from_slice(slice);
```
Will need to consider fallible allocation here.

baumanj replied:
It looks pretty straightforward to add a `FallibleVec::try_extend_from_slice` to [`mp4parse_fallible`](https://github.com/mozilla/mp4parse_fallible). Would you like me to add that before merging this?

I notice there are a couple other [places where `extend_from_slice` is used](https://github.com/mozilla/mp4parse-rust/search?q=extend_from_slice&unscoped_q=extend_from_slice). Do those also need updating, or is this case fundamentally different (aside from the slice being likely larger)?

kinetiknz replied:
The small ones are fine, but it'd be worth investigating the potential sizes (or source of sizes, e.g. from untrusted data) of some of them.

Generally, if the allocation could be "large" or the size of the allocation is user controlled without limits, use a fallible allocator.  If the allocation is "small", trusted code controls the size, it's a user-supplied size but is known to be small, or it's a situation where continuing after an allocation failure is impossible, it's fine to use the infallible allocators.  The concept of "large" and "small" are somewhat nebulous and there are edge cases to consider (e.g. if you've got a large existing allocation and try to grow it by a small amount, it can fail), so it's not entirely straightforward.

I tried to find good documentation on Gecko's policy (which we try to follow here), but only found https://developer.mozilla.org/en-US/docs/Mozilla/Infallible_memory_allocation#Choosing_a_memory_allocator which is fairly terse.

baumanj replied:
It looks like there are 4 existing instances:
1. [`pssh_box.extend_from_slice(b"pssh");`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1049): definitely small
1. [`pssh_data.extend_from_slice(pssh.system_id.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1383): in practice, [exactly 16 bytes](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1030)
1. [`pssh_data.extend_from_slice(data_len.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1384): [4 bytes](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1380)
1. [`pssh_data.extend_from_slice(pssh.box_content.as_slice());`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L1385): doesn't seem to be inherently bounded; [size based on untrusted input](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L1022).

I'll just change the last one and add a comment.

SingingTree commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1173,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+        // which has no `construction_method` field. It does say:
+        // "For maximum compatibility, version 0 of this box should be used in preference to
+        //  version 1 with `construction_method==0`, or version 2 when possible."
+        // We take this to imply version 0 can be interpreted as using file offsets.
```
That is my reading too, and I've been trying to figure out if there's a rule somewhere the makes this explicit. For example if there is a blanket rule with the ISOBMFF where a field that does not exist in for a given version of a box is treated as 0, which would be annoying in it's own way, but would at least give explicit clarity to this case :|

baumanj commented:
Thanks for all the feedback, @kinetiknz! I'll be addressing all the issues and update the PR soon. To reply to some of the higher level comments:

> Would it make sense to exclude mp4parse/tests/avif and use a git submodule of https://github.com/AOMediaCodec/av1-avif instead?

Ah, good idea. I'll look into that.

> New C API needs a new fuzzer added in mp4parse_capi/fuzz/fuzz_targets

Will do.

> I'm not sure about read_avif_meta diverging from read_meta and generally splitting the parser implementation too much to support AVIF; we should be aiming to share as much of the code as makes sense

Yeah, I was hoping to share more code as well, but I didn't see a way that made sense. Since HEIF so so different from video, nothing in [`MediaContext`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L677) would've been useful to AVIF. I suppose we could make dramatic changes to [`read_mp4`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L923) and [`read_meta`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2569) so that they're shared, but conceptually they have nothing really in common, so it didn't seem like a win in terms of complexity of maintainability. Also, it seemed especially disadvantageous to perturb code that is working well for the video case in order to add functionality that seems basically orthogonal.

If you see a different approach that would allow more code to be shared or reduce complexity, I'm all for it.

> I need to take a closer pass through the spec(s) to understand why we need the extent stuff

The `iloc` box is definitely at the heart of it, but it was probably the part of the spec (of what I needed to implement) that I found the most confusing.

baumanj commented on mp4parse/src/boxes.rs:
```diff
@@ -94,6 +94,11 @@ impl fmt::Display for FourCC {
+    MediaDataBox                      0x6d64_6174, // "mdat"
+    PrimaryItemBox                    0x7069_746d, // "pitm"
+    ItemInfoBox                       0x6969_6e66, // "iinf"
+    ItemInfoEntry                     0x696e_6665, // "infe"
```
I was matching the names used in 14496-12:2015 § 8.11.6.2 (the syntax section):
![Screen Shot 2020-02-26 at 12 43 13 PM](https://user-images.githubusercontent.com/5906042/75386066-ded39100-5895-11ea-99fb-f9c5c6faf672.png)
![Screen Shot 2020-02-26 at 12 43 02 PM](https://user-images.githubusercontent.com/5906042/75386068-df6c2780-5895-11ea-8c5d-33d10d44fa5c.png)

`iinf` doesn't have it's own section, so the only reference to its full name is the syntax section for `infe`, which uses `ItemInfoBox`. The name `ItemInformationBox` doesn't appear in the spec.

This is contrary to the existing code's [`minf`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/boxes.rs#L106) and [`schi`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/boxes.rs#L136) whose names in the section heading and syntax section both use `Information` rather than `Info`:
![Screen Shot 2020-02-26 at 12 47 53 PM](https://user-images.githubusercontent.com/5906042/75386752-11ca5480-5897-11ea-94e3-65e5f3c432ff.png)
![Screen Shot 2020-02-26 at 12 48 16 PM](https://user-images.githubusercontent.com/5906042/75386795-260e5180-5897-11ea-8d97-9ee660e3d28b.png)

[`sinf`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/boxes.rs#L138), on the other hand uses "Protection Scheme Information Box" as the title of § 8.12.1, but `ProtectionSchemeInfoBox` in the syntax section:
![Screen Shot 2020-02-26 at 12 48 38 PM](https://user-images.githubusercontent.com/5906042/75386828-358d9a80-5897-11ea-874b-9a4cab0dc470.png)

It's frustrating that the spec isn't more internally consistent, but which would we rather use as normative? And in the case that a box name is only present in the syntax section, what then?

Related: I'm not sure where the [`itif`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/boxes.rs#L152) definition comes from. I don't see it in any of the ISO specs I have, nor does any of the other code seem to reference `itif` or `MetadataItemInformationBox`. Is it necessary?

kinetiknz replied:
My mistake, let's stick with the names you've got.  I was looking at the section titles and assuming the spec was consistent between those and the syntax sections.  We should be using the names from the syntax sections where we can.

Looks like `itif` is snuck in with the iTunes metadata patches but is completely unused, we should remove it.  Good catch.

kinetiknz commented:
> Yeah, I was hoping to share more code as well, but I didn't see a way that made sense. Since HEIF so so different from video, nothing in [`MediaContext`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L677) would've been useful to AVIF. I suppose we could make dramatic changes to [`read_mp4`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L923) and [`read_meta`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2569) so that they're shared, but conceptually they have nothing really in common, so it didn't seem like a win in terms of complexity of maintainability. Also, it seemed especially disadvantageous to perturb code that is working well for the video case in order to add functionality that seems basically orthogonal.

Makes sense to me, thanks.  We can always consider merging/sharing code later, if it makes sense, as the AVIF support grows.

baumanj commented on mp4parse/src/lib.rs:
```diff
@@ -893,6 +1139,17 @@ fn read_fullbox_extra<T: ReadBytesExt>(src: &mut T) -> Result<(u8, u32)> {
 }
 
+// Parse the extra fields for a full box whose flag fields must be zero.
+fn read_fullbox_version_no_flags<T: ReadBytesExt>(src: &mut T) -> Result<u8> {
```
The idea here was that there were several cases (4 just in the code that I added) where it wasn't that the caller didn't _care_ about flags, but that the spec required they be 0. So rather than repeat
```rust
    if flags != 0 {
        return Err(Error::Unsupported("expected flags to be 0"));
    }
```
4 times, or not enforce the flag requirement, it seemed worth packaging that in a function.

If you think it would be better to not enforce that requirement and truly ignore the flags, I'm happy to make that change; just let me know.

baumanj commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1173,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+                // Store any remaining data for potential later extraction
+                if b.bytes_left() > 0 {
+                    let offset = b.offset();
+                    let mut data = Vec::with_capacity(b.bytes_left().try_into()?);
```
Why does [`allocate_read_buf`](allocate_read_buf) explicitly zero the memory as opposed to just allocating the additional capacity? For [the single caller](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2726), it seems to incur an unnecessary cost.

I have other `Vec::with_capacity` callers that aren't `Vec<u8>`, so I'll need to make some change here. What's your preference?
1. Modify `allocate_read_buf` to take a type parameter and not fill its capacity with zeroes
1. Leave `allocate_read_buf` as is, but add a new function like
    ```rust
    fn vec_with_capacity<T>(capacity: usize) -> std::result::Result<Vec<T>, ()> {
        #[cfg(feature = "mp4parse_fallible")]
        {
            let mut v = Vec::new();
            FallibleVec::try_reserve(&mut v, capacity)?;
            Ok(v)
        }
        #[cfg(not(feature = "mp4parse_fallible"))]
        {
            Ok(Vec::with_capacity(capacity))
        }
    }
    ```
1. Something else
?

kinetiknz replied:
The zeroing is due to https://github.com/mozilla/mp4parse-rust/issues/172

baumanj replied:
Ooh, that's subtle. I'm going to add a comment back to the issue for posterity, if you don't mind.

Do you have any preference regarding whether or not to add `vec_with_capacity` versus modifying `allocate_read_buf` to handle types other than `u8`?

If I implement `vec_with_capacity` as described above, I don't think we need to worry about the issue in https://github.com/mozilla/mp4parse-rust/issues/172 since it doesn't provide any access to potentially uninitialized memory the way [the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L70) did. It gets around that by using [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end) which takes a `&mut Vec<u8>` `buf` rather than [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read), which takes a `&mut [u8]` `buf`.

It occurs to me that we could eliminate the need for zeroing out the memory in `allocate_read_buf` (and for `allocate_read_buf` ifself) if we switched [the code here](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse/src/lib.rs#L2726-L2727) from using `read` to `read_to_end` with `take` like so:

```rust
    if let Ok(mut buf) = allocate_read_buf(size) {
        let r = src.read(&mut buf)?;
```
becomes
```rust
    if let Ok(mut buf) = vec_with_capacity(size) {
        let r = src.take(size.to_u64()).read_to_end(&mut buf)?;
```

The only caveat I can see is that if the `Vec` passed to `read_to_end` didn't already have sufficient capacity, it could end up doing infallible allocation. But that shouldn't be an issue as long as we use `vec_with_capacity`.

Thoughts?

kinetiknz replied:
> Ooh, that's subtle. I'm going to add a comment back to the issue for posterity, if you don't mind.

Good idea, thanks.

> Do you have any preference regarding whether or not to add `vec_with_capacity` versus modifying `allocate_read_buf` to handle types other than `u8`?

I think we only need one, so making it handle other simple types makes sense.  I don't mind which one we keep but vec_with_capacity is probably a better name.

> If I implement `vec_with_capacity` as described above, I don't think we need to worry about the issue in #172 since it doesn't provide any access to potentially uninitialized memory the way [the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L70) did. It gets around that by using [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end) which takes a `&mut Vec<u8>` `buf` rather than [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read), which takes a `&mut [u8]` `buf`.

I think you still have the same issue, in that you can't pass a Vec with an uninitialized backing store to Read::read_to_end because the trait doesn't promise implementations won't read from the &mut Vec<u8> before writing to it.  Although the docs only include a warning for Read::read, it's actually a restriction on the entire trait.  Once Read::initializer is stable, we can use that to signal that our implementation of Read doesn't read from the buffers passed in.

baumanj replied:
Maybe I'm not understanding https://github.com/mozilla/mp4parse-rust/issues/172 correctly. Can you let me know where my logic is flawed?

As I understand it, the unsoundness came from [this line in the old version of `allocate_read_buf`](https://github.com/mozilla/mp4parse-rust/blob/26e614bbcb4d7322dc2e4b7e15391bdb30b9f7be/mp4parse/src/lib.rs#L75):
```rust
        unsafe { buf.set_len(size); }
```
because it violated [the safety requirement of `Vec::set_len()`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html?search=#safety-1):
> The elements at `old_len..new_len` must be initialized.

And as mentioned in [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read):
> It is your responsibility to make sure that `buf` is initialized before calling `read`. Calling `read` with an uninitialized `buf` (of the kind one obtains via `MaybeUninit<T>`) is not safe, and can lead to undefined behavior.

So, because `allocate_read_buf` called `set_len` without initializing the additional capacity, it was uninitialized and the call to `Read::read` was unsound. More generally, `allocate_read_buf` returned a vector that was invalid because uninitialized memory was now accessible via the safe interface.

Hovever, I don't see how the `vec_with_capacity` case has the same issue. [There's no `unsafe` code in its implementation](https://github.com/mozilla/mp4parse-rust/blob/cd8b3d7c6663a8866ab4a891749e54b809257ebf/mp4parse/src/lib.rs#L144-L146), so how does it introduce any uninitialized memory into the memory of the vector's elements? The only potentially uninitialized memory should be in the internal buffer beyond the valid length of the vector.

Take [this example code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=060c6f69872ebd39e69d585effc49063):

```rust
    let mut reader: &[u8] = b"1234567890";
    let mut buf = Vec::with_capacity(5);
    reader.read(&mut buf).unwrap();
```

After the first call to `reader.read`, `buf` will be empy. Nothing will be read into it because despite having a capacity of 10, it has a length of 0. Contrast that with [identical code except replacing `read` with `read_to_end`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=576c38ad9c0555df7ac4bfcc33587695), and the result is that `buf` contains all 10 elements.

I think the difference is easy to miss because in both cases we pass `&mut buf`, but in the `read` call, that is auto derefed to a `&mut [u8]` slice, and in the `read_to_end` case, it remains a `&mut Vec<u8>`. The callee can't change the length of the prior, but it can change the length of the latter, which is why [`Read::read_to_end`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_to_end)'s documentation describes it as "append[ing] data to `buf`, whereas [`Read::read`](https://doc.rust-lang.org/stable/std/io/trait.Read.html#tymethod.read)'s says it "pull[s] some bytes from this source into the specified buffer".

Unless I'm missing something, we have the opportunity to avoid the cost of zeroing out the buffer (as with the earlier version of `allocate_read_buf`), but eliminating the potential for unsoundness. Does that seem worth it?

kinetiknz replied:
I'm confusing things by thinking only about the allocate_read_buf impl. I agree with your comment (thanks for the detailed reasoning!), so replacing  allocate_read_buf impl + read with vec_with_capacity + read_to_end allows us to eliminate the extra initialization safely.

baumanj replied:
I ended up taking a slightly different approach from what I suggested in https://github.com/mozilla/mp4parse-rust/pull/193#discussion_r386771240 since I realized I could simplify things by making [our new `read_to_end` with fallible allocation support](https://github.com/mozilla/mp4parse-rust/blob/avif-pitm/mp4parse/src/lib.rs#L168) take care of doing all the allocation.

The heart of the change is [here](https://github.com/mozilla/mp4parse-rust/pull/193/commits/9246ad69e75f93644816b9a31ac3c90eb20cd92d). I ended up splitting out some other cleanup-type changes into [the previous commit](https://github.com/mozilla/mp4parse-rust/pull/193/commits/ba1b7c94c9d2d2e7f39019b9ea9cbc532b810d90).

@kinetiknz, @SingingTree: could you please give this one more look before I merge?

baumanj commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1173,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+                    let offset = b.offset();
+                    let mut data = Vec::with_capacity(b.bytes_left().try_into()?);
+                    let _bytes_read = b.read_to_end(&mut data)?;
+                    mdats.push(Mdat { offset, data });
```
I'll update all the `Vec::push` calls I added.

Should the existing code [`audio_sample_infos.push(sample_info);`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L734) and [`video_sample_infos.push(sample_info);`](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/src/lib.rs#L872) also be converted to use `vec_push()`?

baumanj replied:
I've tentatively changed those as well:
- [`video_sample_infos`](https://github.com/mozilla/mp4parse-rust/pull/193/files#diff-2c323c51cd6859bdc9990f2c461fe517R984)
- [`audio_sample_infos`](https://github.com/mozilla/mp4parse-rust/pull/193/files#diff-2c323c51cd6859bdc9990f2c461fe517R843)

Let me know if that looks good, or you'd rather I revert them or do something else.

baumanj commented on mp4parse_capi/src/lib.rs:
```diff
@@ -437,6 +506,45 @@ pub unsafe extern fn mp4parse_read(parser: *mut Mp4parseParser) -> Mp4parseStatu
+    match r {
+        Ok(_) => Mp4parseStatus::Ok,
+        Err(Error::NoMoov) | Err(Error::InvalidData(_)) => { // Add Error::NoPitm or something
+            // Block further calls. We've probable lost sync.
```
Ah, copy pasta. Will fix both instances.

baumanj commented on mp4parse_capi/src/lib.rs:
```diff
@@ -888,6 +996,33 @@ pub unsafe extern fn mp4parse_get_track_video_info(parser: *mut Mp4parseParser,
+
+    let context = (*parser).context();
+
+    // maybe a check for a validly parsed context?
```
Will do. I'll also file issues for tracking purposes.

baumanj commented:
This will be busted until https://github.com/mozilla/mp4parse_fallible/pull/6 is merged (presumably requiring a new version and so one final update here), but most of the other issues should be addressed, so I thought it was worth getting the review cycle started again.

baumanj commented:
I note that the CI didn't fail, which leads me to believe that it doesn't test building with `--features mp4parse_fallible`. Shall I add an issue to rectify that?

SingingTree commented on mp4parse/tests/public.rs:
```diff
@@ -594,3 +599,42 @@ fn public_video_av1() {
+    let context = &mut mp4::AvifContext::new();
+    let input = &mut File::open(IMAGE_AVIF_GRID).expect("Unknown file");
+    mp4::read_avif(input, context).expect("read_avif failed");
+    // Add some additional checks
```
Re comment, do we want any additional checks before merging this, or will we handle this case when we un-ignore the test?

baumanj replied:
I think the nature of the additional checks will be more evident once we add the support for non-av01 primary items. I've added https://github.com/mozilla/mp4parse-rust/issues/198 to track that, and will add a link to the code here.

SingingTree commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1227,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+        check_parser_state!(b.content);
+    }
+
+    let primary_item_id = primary_item_id.ok_or(Error::InvalidData("primary_item_id not present in iinf box"))?;
```
Is this error case triggered by the `pitm` box being absent? If so, could we change the error string?

baumanj replied:
Good catch! Will change to `"Required pitm box not present in meta box"`.

SingingTree commented on mp4parse/src/lib.rs:
```diff
@@ -916,6 +1227,338 @@ fn skip_box_remain<T: Read>(src: &mut BMFFBox<T>) -> Result<()> {
+        let base_offset = iloc.read_u64(base_offset_size.to_bits())?;
+        let extent_count = iloc.read_u16(16)?;
+
+        debug_assert!(extent_count >= 1, "extent_count must have a value 1 or greater per ISO 14496-12:2015 § 8.11.3.3");
```
I suggest we just explicitly handle this as this invariant can be violated by malformed files. I think the fuzzers are likely to run into this sooner or later.

kinetiknz commented:
I see the CI with --all-features is failing to build afl.rs.  Rather than any spend time fixing it, I think we can remove anything related to afl.rs fuzzing now, since we're using the cargo fuzz infrastructure under mp4parse_capi/fuzz, making the older fuzzing support obsolete.

baumanj commented:
I think the issue is that `abort_on_panic` was missing from [`mp4parse_capi/Cargo.toml`](https://github.com/mozilla/mp4parse-rust/blob/master/mp4parse_capi/Cargo.toml). However, I do like deleting code, so I'll try removing all the afl stuff and see if CI passes.

baumanj commented:
Well, that seemed to solve the `afl.rs` issue, but now travis Windows jobs with `--all-features` are failing in a non-obvious way. Any ideas, @kinetiknz? Some quick googling makes me think we could be running into https://github.com/rust-lang/rust/issues/39487#issuecomment-277401239.

kinetiknz commented:
The tests are crashing with STATUS_HEAP_CORRUPTION.  This is because it's not safe to build with mp4parse_fallible enabled unless it's also guaranteed that the system allocator matches the Rust allocator (which is guaranteed within Gecko).  That's briefly mentioned [here](https://github.com/mozilla/mp4parse-rust/blob/bbb4a1217d2826e95b0a843e73c9e1a2faa23953/mp4parse_capi/Cargo.toml#L38).

Since Rust dropped jemalloc, we're getting lucky on the other Travis platforms since the system and Rust allocators happen to match.  Windows has different allocator rules, so we hit this issue.

I think it's fine to run tests without mp4parse_fallible built on Travis, since we still get coverage via Gecko's testsuite.

baumanj commented:
> I think it's fine to run tests without mp4parse_fallible built on Travis, since we still get coverage via Gecko's testsuite.

Ah, that makes sense. Thanks for the explanation. In the spirit of getting a little bit of early warning, how about just running `cargo check --all --tests --all-features` and only running `cargo test` without `mp4parse_fallible`?

baumanj commented:
This is kind of odd. Investigating [the CI failures](https://travis-ci.org/mozilla/mp4parse-rust/builds/657593327?utm_source=github_status&utm_medium=notification) following [adding `cargo check`](https://github.com/mozilla/mp4parse-rust/pull/193/commits/4cba7f5fc576d60a68ce7c8a033330ceacb79fad), I can reproduce this consistently locally (macOS):
```
$ cargo clean && cargo check && cargo test --all
[…]
test build_ffi_test ... FAILED

failures:

---- build_ffi_test stdout ----
status: exit code: 2
--- stdout ---
rm -f test
rm -f *.a.out
rm -f *.o *.a
rm -f *.d
rustc -g --crate-type staticlib --crate-name mp4parse \
	  --emit dep-info,link=libmp4parse.a \
	  --print native-static-libs \
	  -L /Users/jbauman/src/mp4parse-rust_m-u-current/target/debug/build/mp4parse_capi-45f37232212af4dc/out/../../../deps ../src/lib.rs \
	  2> libmp4parse.a.out || cat libmp4parse.a.out >&2
c++ -g -Wall -std=c++11 -I../include/ -c test.cc
c++ -g -Wall -std=c++11 -I../include/ -o test *.o libmp4parse.a 

-- stderr ---
error[E0464]: multiple matching crates for `mp4parse`
  --> ../src/lib.rs:37:1
   |
37 | extern crate mp4parse;
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidates:
           crate `mp4parse`: /Users/jbauman/src/mp4parse-rust_m-u-current/target/debug/deps/libmp4parse-d0bac2999dfaf13d.rlib

error[E0463]: can't find crate for `mp4parse`
  --> ../src/lib.rs:37:1
   |
37 | extern crate mp4parse;
   | ^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0463`.
clang: error: no such file or directory: 'libmp4parse.a'
make: *** [test] Error 1

thread 'build_ffi_test' panicked at 'assertion failed: output.status.success()', mp4parse_capi/tests/build_ffi_test.rs:19:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```
however both
```
$ cargo clean && cargo test --all && cargo test --all
```
```
$ cargo clean && cargo build && cargo test --all
```
work fine

So, it's not about features, or about running things multiple times, but something that `cargo check` is doing is causing that failure. I'll investigate.

kinetiknz commented:
build_ffi_test is pretty wacky, sorry. It's trying to link the built mp4parse, so it relies on output from the build phase.  The test phase produces a different path/binary, so the makefile fails to find the input.  (Re)writing it now, it'd make more sense to use cc-rs instead of shelling out to make, and there's probably a better way to find the path of the link target than hardcoding the makefile.

It'd be great to fix it but we can fix it independently from your PR, so don't let it block your progress on merging if you'd prefer to focus on that first.

baumanj commented:
> build_ffi_test is pretty wacky, sorry. It's trying to link the built mp4parse, so it relies on output from the build phase.

Requiring output from the build phase isn't that surprising to me, but what I'm not really grokking is why adding in a `cargo check` beforehand would break things. Even `cargo check && cargo build && cargo test` fails, so it's not the absence of `cargo build` that's causing the failure.

In any case, since this is definitely a detour relative to the purpose of this PR, and you seem to be more familiar with the code in question, I'll leave it alone. In the meantime, I think a good workaround would be putting the `cargo check --all --verbose $RELEASE --tests --all-features` _after_ the `cargo test` in CI. That should hopefully still get us some checking of `mp4parse_fallible`-feature code.

I'll try that change shortly. And if that doesn't work, I'll just give up on changing the CI to include feature checking for now.
